master/static/srv/ship_logs.py

```diff
@@ -0,0 +1,567 @@
+determined library.  The reason is that python virtual env errors in user containers are common
+scenarios for users to encounter, and we must be able to still ship logs in those scenarios.
+
+The only thing that is allowed to break the log shipper is a misconfigured cluster (if the log
+shipper isn't able to connect to the master) or if python isn't installed.
+
+But if you are thinking of making this log shipper do anything other than ship logs, or if you are
+thinking of adding any dependencies not inside the standard library, stop.  Let the log shipper
+just ship logs; it's too important of a job to mix it with anything else.
+"""
+
+import datetime
+import io
+import json
+import logging
+import os
+import queue
+import re
+import signal
+import ssl
+import subprocess
+import sys
+import threading
+import time
+import traceback
+import urllib.request
+from typing import IO, Any, Dict, List, Optional, cast
+
+# Duplicated from determined/__init__.py.  It's nice to keep them in sync.
+LOG_FORMAT = "%(levelname)s: [%(process)s] %(name)s: %(message)s"
+
+
+# Example log message given below.
+# 2022-05-12 16:32:48,757:gc_checkpoints: [rank=0] INFO: Determined checkpoint GC, ...
+# Below regex is used to extract the rank field from the log message.
+# Excluding empty spaces this regex matches rank in the above example as [rank=0]
+rank = re.compile(r"(?P<space1> ?)\[rank=(?P<rank_id>([0-9]+))\](?P<space2> ?)(?P<log>.*)")
+# Below regex is used to extract the message severity from the log message.
+# Excluding empty spaces and delimiter(:) this regex matches message severity level in the above
+# example as INFO
+level = re.compile(
+    r"(?P<space1> ?)(?P<level>(DEBUG|INFO|WARNING|ERROR|CRITICAL)):(?P<space2> ?)(?P<log>.*)"
+)
+lineend = re.compile(rb"[\r\n]")
+
+
+# Interval at which to force a flush.
+SHIPPER_FLUSH_INTERVAL = 1  # How often to make API calls
+
+# Full jitter time on encountering an API exception.
+SHIPPER_FAILURE_BACKOFF_SECONDS = 1
+
+# Max size of the log buffer before forcing a flush.
+LOG_BATCH_MAX_SIZE = 1000
+
+# Max size of the shipping queue before we start to apply backpressure by blocking sends. We would
+# only hit this if we got underwater by three full batches while trying to ship a batch.
+SHIP_QUEUE_MAX_SIZE = 3 * LOG_BATCH_MAX_SIZE
+
+
+def read_newlines_or_carriage_returns(fd):
+    r"""
+    Implement our own line buffering.  Unlike the default io.BufferedReader used in
+    subprocess.Popen(bufsize=-1), we read until we encounter either '\n' or \r', and treat that as
+    one line.
+    """
+    # Ship lines of length of DEFAULT_BUFFER_SIZE, including the terminating newline.
+    limit = io.DEFAULT_BUFFER_SIZE - 1
+    nread = 0
+    chunks = []
+
+    def oneline():
+        nonlocal nread
+        nonlocal chunks
+        out = b"".join(chunks).decode("utf8")
+        chunks = []
+        nread = 0
+        return out
+
+    while True:
+        buf = fd.read(limit - nread)
```

Comment:
I'm also not familiar enough with operations here, too. Maybe it depends on what type exactly `fd` is, but when this gets to the end of a stream that's been consumed up to the end but is still open, does it block? Talking about inputs and outputs in the docstring would help. Maybe typing would be sufficient.

Reply:
Yes, it blocks.  Roughly speaking, that's why we're using threads here.

We could make it a non-blocking filedescriptor and use selectors but I wasn't sure if that would exacerbate the raciness of stdout and stderr log capturing.

Reply:
Thanks! I get this function, now.